### ISSUE TITLE
Services: switching to details page is ignored sometimes

### ISSUE THREAD

**Issue by martinpitt** [cockpit-project/cockpit#18443]
I found these a lot while working on #18422, but this already [happens on main](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18442-20230303-083420-6f3008ed-ubuntu-stable/log.html#297). 

It's a really tough one. It sometimes repro locally, but of course never when I want it to. So I [added some debugging](https://github.com/cockpit-project/cockpit/pull/18422/commits/3f6b1c8a9e82f219f1d9b7d3db6b1edf43a323fd).

I was originally hoping that `this.props.path` is correct, and we just have deferred handlers which hardcode going to `[]`. But this is not correct. E.g. [here](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18422-20230303-075148-3f6b1c8a-rhel-8-8/log.html#328-2) we have:
```
-> ph_mouse("tr[data-goto-unit=\"condtest.service\"] a:not([disabled]):not([aria-disabled=true])","click",0,0,0,false,false,false,false)
> debug: locationchanged {"type":"service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> debug: XXX ServicesPageBody render list for path []
> debug: componentDidUpdate change detected, recomputing selected units
> debug: XXX ServicesPageBody render list for path []
> debug: locationchanged {"type":"service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> debug: XXX ServicesPageBody render list for path []
> debug: componentDidUpdate change detected, recomputing selected units
> debug: XXX ServicesPageBody render list for path []
> debug: XXX ServicesPageBody render list for path []
> log: YYY Clicked on service in list, going to condtest.service
<- {'type': 'undefined'}
```
:arrow_up:  this does not seem to have *any* effect.

```
-> wait: ph_collected_text_is("#statuses .status","Not runningAutomatically starts")
> log: YYY ServicesPageBody onOptionsChanged {"activestate":"[]","filestate":"[]","name":""} calling go to services list; path prop []
> debug: locationchanged {"type":"service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> debug: XXX ServicesPageBody render list for path []
> debug: componentDidUpdate change detected, recomputing selected units
> debug: XXX ServicesPageBody render list for path []
> debug: locationchanged {"type":"service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
> debug: XXX ServicesPageBody render list for path []
> debug: componentDidUpdate change detected, recomputing selected units
> debug: XXX ServicesPageBody render list for path []
> log: YYY ServicesPageBody onOptionsChanged {"activestate":"[]","filestate":"[]","name":""} calling go to services list; path prop []
> debug: XXX ServicesPageBody render list for path []
```
I was hoping that theonOptionsChanged handler was running too late, and at that point we already had `this.props.path == ["condtest.service"]`, but that's not the case.

In the same vein, this [testBasic failure](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18422-20230303-075148-3f6b1c8a-fedora-37-devel/log.html#326) also completely ignores the URL change:
```
-> ph_go("/system/services#/nonexistent")    
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{},"href":"/"}    
> debug: XXX ServicesPageBody render list for path []    
> log: YYY ServicesPageBody onOptionsChanged {"activestate":"[]","filestate":"[]","name":""} calling go to services list; path prop []
> debug: XXX ServicesPageBody render list for path []    

> log: YYY ServicesPageBody onOptionsChanged {"activestate":"[]","filestate":"[]","name":""} calling go to services list; path prop []
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{},"href":"/"}    
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{},"href":"/"}    
> debug: XXX ServicesPageBody render list for path []    
> debug: XXX ServicesPageBody render list for path []    
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{},"href":"/"}    
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{},"href":"/"}    
> debug: XXX ServicesPageBody render list for path []    
> debug: XXX ServicesPageBody render list for path []    
```
So we get a zillion completely unnecessary usePageLocation triggers, but  they are all wrong :cry:

_Originally posted by @martinpitt in https://github.com/cockpit-project/cockpit/issues/18422#issuecomment-1453043638_

**Comment by martinpitt** [comment]
@mvollmer says he remembers something similar when we investigated the broken back button in #13091. Might be worth looking at the issue history and PR #16938. @KKoukiou says she is seeing the same thing on the Users page (which also uses `useLocation`). There's apparently something wrong with `cockpit.location` and how browsers handle this.

**Comment by martinpitt** [comment]
With some more debugging, I get [this](https://cockpit-logs.us-east-1.linodeobjects.com/pull-18422-20230303-165501-04bdc094-centos-8-stream-pybridge/log.html#328-2):
```
> debug: XXX ServicesPageBody render list for path []
> log: YYY Clicked on service in list, going to condtest.service
> log: XXX cockpit.location.get; window.location.hash "#/?type=service" unchanged last_loc
> log: XXX cockpit.location.get; window.location.hash "#/?type=service" unchanged last_loc
> log: YYY cockpit.Location.go: changed window.location.hash to #/condtest.service?type=service
<- {'type': 'undefined'}
-> wait: ph_collected_text_is("#statuses .status","Not runningAutomatically starts")
> log: XXX cockpit.location.get; window.location.hash "#/condtest.service?type=service" creating new last_loc
> log: XXX cockpit.location.get; window.location.hash "#/condtest.service?type=service" unchanged last_loc
> log: YYY ServicesPageBody onOptionsChanged; cockpit.location {"url_root":"","path":["condtest.service"],"options":{"type":"service"},"href":"/condtest.service?type=service"} merged options {"type":"service"} calling go to services list; path prop []
> log: XXX cockpit.location.get; window.location.hash "#/condtest.service?type=service" unchanged last_loc
> log: YYY cockpit.Location.go: changed window.location.hash to #/?type=service
> log: XXX cockpit.location.get; window.location.hash "#/?type=service" creating new last_loc
> debug: locationchanged {"type":"service"}
> log: XXX cockpit.location.get; window.location.hash "#/?type=service" unchanged last_loc
> log: XXX usePageLocation changed {"url_root":"","path":[],"options":{"type":"service"},"href":"/?type=service"}
```
I.e. it seems like the "Clicked" actually *did* work, but then the `onOptionsChanged` handler fired and reset the path back to `[]`. The ServicesPageBody did not even see the `path` property change yet, so we can't just replace `[]` with `this.props.path`. So the async onOptionsChanged handler (which is called by `ServicesPageFilters`) steps on its own toes, and the `locationchanged` handler is way too laggy to even see the intermediate state.

**Comment by martinpitt** [comment]
Heureka! I've got it, initial coarse fix is in #18422. I'll clean it up and refine it a bit.
